File: Диплом/Программа/MFDES.py
import numpy as np
import scipy.linalg
import math
import matplotlib.pyplot as plt
from matplotlib import cm
from scipy.special import gamma as GAMMA, factorial, beta as BETA
from FDESBase import *

# ModifiedFractionalDifferentialEquationSolver
class MFDES(FDESBase):
    
    def __init__(self, L, R, T, alpha, beta, gamma, D, h, tau, psi, f, phiL=None, phiR=None):
        super().__init__(L, R, T, alpha, beta, gamma, D, h, tau, psi, f, phiL, phiR)
        
    def solve(self):
        
        for t in range(1, self.k + 1):
            
            # Заполняем вектор d
            d = np.array([0.0 for i in range(self.n + 1)])
        
            for i in range(self.n + 1):
                for j in range(1, t + 1):
                    d[i] += self.g(self.gamma, j) * self.result[t-j][i]

                d[i] = d[i] - (self.tau**self.gamma) * self.f(self.x_i(i), self.t_k(t))
            
            # Заполняем матрицу А
            for i in range(self.n + 1):
                for j in range(self.n + 1):
                    if i == j:
                        self.A[i][j] = self.a(self.x_i(i), self.t_k(t)) * self.g(self.alpha, 1.0) + \
                                       self.b(self.x_i(i), self.t_k(t)) * self.g(self.alpha, 1.0) - 1.0
                    elif i + 1 < j:
                        self.A[i][j] = self.b(self.x_i(i), self.t_k(t)) * self.g(self.alpha, j - i + 1.0)
                    elif i > j + 1:
                        self.A[i][j] = self.a(self.x_i(i), self.t_k(t)) * self.g(self.alpha, i - j + 1.0)
                    elif i == j + 1:
                        self.A[i][j] = self.a(self.x_i(i), self.t_k(t)) * self.g(self.alpha, 2.0) + \
                                       self.b(self.x_i(i), self.t_k(t))
                    else:
                        self.A[i][j] = self.a(self.x_i(i), self.t_k(t)) + \
                                       self.b(self.x_i(i), self.t_k(t)) * self.g(self.alpha, 2.0)
            
            # Если есть граничные условия
            if not(self.phiL == None) and not(self.phiR == None):
                d[0] = self.phiL(self.t_k(t))
                d[-1] = self.phiR(self.t_k(t))
                
                for i in range(self.n + 1):
                    self.A[0][i] = 1.0 if i == 0 else 0.0
                    self.A[self.n][i] = 1.0 if i == self.n else 0.0
            
            # Решаем систему
            self.result[t] = scipy.linalg.solve(self.A, d)
            
            # Если есть граничные условия
            if not(self.phiL == None) and not(self.phiR == None):
                self.result[t][0] = self.phiL(self.t_k(t))
                self.result[t][-1] = self.phiR(self.t_k(t)) 
            
            
        return self

File: Диплом/Программа/diplom/lib/MFDES.cpp
#include "MFDES.h"

MFDES::MFDES(const FDESbase& _p) : FDESbase(_p) {}

void MFDES::solve() {

    for (ull i = 0; i <= n; i++) {
        result[0][i] = psi(x_i(i));
    }

    Matrix A(n + 1);

    // create matrix A
    for (ull i = 0; i <= n; i++) {
        for (ull j = 0; j <= n; j++) {
            if (i == j) {
                A[i][j] = a(x_i(i)) * g(alpha, 1) + b(x_i(i)) * g(alpha, 1) - 1.0;
            } else if (i + 1 < j) {
                A[i][j] = b(x_i(i)) * g(alpha, j - i + 1);
            } else if (i > j + 1) {
                A[i][j] = a(x_i(i)) * g(alpha, i - j + 1);
            } else if (i == j + 1) {
                A[i][j] = a(x_i(i)) * g(alpha, 2) + b(x_i(i)) - c(x_i(i));
            } else {
                A[i][j] = a(x_i(i)) + b(x_i(i)) * g(alpha, 2) + c(x_i(i));
            } 
        }
    }

    //// 1-border conditions
    if (is_borders) {
        for (ull i = 0; i <= n; i++) {
            A[0][i] = 0.0;
            A[n][i] = 0.0;
        }

        A[0][0] = 1.0;
        A[n][n] = 1.0;
    }
    
    for (ull t = 1; t <= k; t++) {
        
        // create d-vector
        std::vector<double> d(n + 1, 0.0);

        for (ull i = 0; i <= n; i++) {
            for (ull j = 1; j <= t; j++) {
                d[i] += g(gamma, j) * (result[t-j][i] - result[0][i]);
            }
            d[i] -= result[0][i];
            d[i] -= std::pow(tau, gamma) * f(x_i(i), t_k(t));
        }

        //// borders
        if (is_borders) {
            d[0] = phiL(t_k(t));
            d[n] = phiR(t_k(t));
        }
        
        // solve system
        auto r = A.Solve(d);
        
        for (ull i = 0; i < r.size(); i++) {
            result[t][i] = r[i];
        }
    }
}